pkg/meta/sql.go

```diff
@@ -3133,155 +3092,95 @@ func (m *dbMeta) LoadMeta(r io.Reader) error {
 			var err error
-			if n, err = s.Insert(beanSlice); err != nil {
-				return err
+			if index == len(chs)-1 { // multiple tables
+				n, err = s.Insert(beans...)
+			} else { // one table only
+				n, err = s.Insert(beans)
+			}
+			if err == nil && int(n) != len(beans) {
+				err = fmt.Errorf("only %d records inserted", n)
 			}
-			return nil
-		})
-		if err != nil {
 			return err
-		} else if d := len(beanSlice) - int(n); d > 0 {
-			return fmt.Errorf("%d records not inserted: %+v", d, beanSlice)
-		}
-		lbar.IncrInt64(n)
-		return nil
+		})
 	}
-
-	addToBatch := func(batch *[]interface{}, bean interface{}) error {
-		*batch = append(*batch, bean)
-		if len(*batch) >= batchSize {
-			if err := insertBatch(*batch); err != nil {
-				return err
+	var wg sync.WaitGroup
+	for i := range chs {
+		chs[i] = make(chan interface{}, batch*2)
+		wg.Add(1)
+		go func(i int) {
+			defer wg.Done()
+			buffer := make([]interface{}, 0, batch)
+			for bean := range chs[i] {
+				buffer = append(buffer, bean)
+				if len(buffer) >= batch {
+					if err := insert(i, buffer); err != nil {
+						logger.Fatalf("Write %d beans in channel %d: %s", len(buffer), i, err)
+					}
+					buffer = buffer[:0]
+				}
 			}
-			*batch = (*batch)[:0]
-		}
-		return nil
+			if len(buffer) > 0 {
+				if err := insert(i, buffer); err != nil {
+					logger.Fatalf("Write %d beans in channel %d: %s", len(buffer), i, err)
+				}
+			}
+		}(i)
 	}
 
-	for bean := range beansCh {
-		switch bean.(type) {
-		case *chunk:
-			if err := addToBatch(&chunkBatch, bean); err != nil {
-				return err
-			}
-		case *edge:
-			if err := addToBatch(&edgeBatch, bean); err != nil {
-				return err
-			}
-		case *xattr:
-			if err := addToBatch(&xattrBatch, bean); err != nil {
-				return err
-			}
-		case *node:
-			if err := addToBatch(&nodeBatch, bean); err != nil {
-				return err
-			}
-		case *chunkRef:
-			if err := addToBatch(&chunkRefBatch, bean); err != nil {
-				return err
-			}
-		default:
-			if err := insertBatch([]interface{}{bean}); err != nil {
-				return err
-			}
-		}
+	dm, counters, parents, refs, err := loadEntries(r, func(e *DumpedEntry) { m.loadEntry(e, chs) })
+	if err != nil {
+		return err
 	}
-	for _, one := range [][]interface{}{chunkBatch, edgeBatch, xattrBatch, nodeBatch, chunkRefBatch} {
-		if len(one) > 0 {
-			if err := insertBatch(one); err != nil {
-				return err
-			}
-		}
+	format, _ := json.MarshalIndent(dm.Setting, "", "")
+	chs[5] <- &setting{"format", string(format)}
+	chs[5] <- &counter{usedSpace, counters.UsedSpace}
+	chs[5] <- &counter{totalInodes, counters.UsedInodes}
+	chs[5] <- &counter{"nextInode", counters.NextInode}
+	chs[5] <- &counter{"nextChunk", counters.NextChunk}
+	chs[5] <- &counter{"nextSession", counters.NextSession}
+	chs[5] <- &counter{"nextTrash", counters.NextTrash}
+	for _, d := range dm.DelFiles {
+		chs[5] <- &delfile{d.Inode, d.Length, d.Expire}
 	}
+	for k, v := range refs {
+		chs[3] <- &chunkRef{k.id, k.size, int(v)}
```

Comment:
we should insert those with ref =1, then update them if ref > 1

Reply:
`chunkRef` is not inserted until in this for loop, so no need to update

Reply:
This could be an issue, if there are too many keys in `refs`.

Reply:
there could be more rows than node